### DISCUSSION TITLE
How do I run an async CLR function efficiently from Jint

### DISCUSSION THREAD

**Discussion by Ghostbird** [sebastienros/jint#1463] [Q&A] [answered]
Note that to the script writer it shouldn't matter whether the code is async or not. I have an engine that runs some javascript expressions. It exposes some CLR methods. The underlying implementation of some of those methods is asynchronous in C#. Currently we simply block on those async calls, but I'd prefer not to.

Is there some way to use `async` `await` at the C# level, or is this impossible at the moment?

**Comment by lahma** [comment]
Previous discussions:

* https://github.com/sebastienros/jint/discussions/935
* https://github.com/sebastienros/jint/discussions/943
* https://github.com/sebastienros/jint/discussions/1081

**Reply by Ghostbird** [reply]
I had already read those partially, but it didn't seem to match my use case exactly. I'll give them a better look.

**Reply by Ghostbird** [reply]
I guess I really don't understand the way Jint works. None of these other discussions seem to match what I want.
The basics of what I want _seem_ simple:

1. I create an engine instance.
2. I add a `ClrFunctionInstance` to it.
3. I call the `ClrFunctionInstance` and extract the result.
4. I discard the engine instance.

I'd like my `ClrFunctionInstance` to be a C# function that performs asynchronous operations. There's no concurrency on the engine instance, and nothing is going to access the engine instance if I temporarily yield the thread.

**Reply by lahma** [reply, marked as answer]
Currently the best option is to call `GetAwaiter().GetResult()` to get the async outcome.

**Reply by Ghostbird** [reply]
Yeah, that's what we do currently, I was wondering if I could improve upon that, but I guess I really don't know enough about this. Thanks for explaining!

**Reply by SebastianStehle** [reply]
The jint code is synchronous, which means you cannot just have a normal method that is asynchronous in C#, but looks synchronous in the script. This would also be my preferred method.

So I think you have 3 options for asynchronous code:

1. Give your users old school methods with a callback. I have followed this path.
2. Use promises and ask you users (or whoever writes the script) to use async await.
3. Just use `.Wait()` in your methods (not in your script). I remember that there is a better alternative, something with AsyncPump or so, but I cannot remember the details.

Sorry, lahma was faster ...

**Reply by Ghostbird** [reply]
Thanks for your reply. We've been using something very similar to [this](https://gist.github.com/rid00z/96b875f096cbc7cbcba8) and it works well. It just left me wondering if there was some way to be less blocking.
